src/protocols/radius/encode.c

```diff
@@ -42,35 +42,34 @@ static ssize_t encode_tlv_hdr(uint8_t *out, size_t outlen,
 			      fr_dict_attr_t const **tlv_stack, unsigned int depth,
 			      fr_cursor_t *cursor, void *encoder_ctx);
 
+static inline bool is_encodable(VALUE_PAIR const *vp)
+{
+	return vp && !vp->da->flags.internal;
```

Comment:
That should really also check for `fr_dict_parent_common()`, in case some manages to add DHCPv4 attributes to a RADIUS packet.

Reply:
Easy enough to add; will do. Also, while looking this afternoon to see who calls fr_dict_parent_common() in general, it looks like src/lib/eap_aka_sim/encode.c has yet another batch of _*encodable() functions. I will give it the same treatment as the ones under protocol/*.

Reply:
eap_aka_sim/encode.c has different logic.  It's inefficient to use fr_dict_parent_common.  Should just take an `fr_dict_t *` and compare that to da->dict.

Reply:
I spoke too soon; RADIUS is at least nonobvious. From the fr_<foo>_ctx_t * that the other encodability checking functions cast the void * they're handed to, you can get to a  fr_dict_t * to compare with da->dict (if that's not the one to compare, please let me know), but fr_radius_ctx_t doesn't  look like it has anything ultimately traceable to a dictionary.

Reply:
You will have to add a `fr_dict_t *dict` to `fr_radius_ctx_t`

there's a `fr_dict_t const *dict_radius;` in `src/protocols/radius/attrs.h`, so you can just initialize it from that:

`packet_ctx = dict_radius`

and that should work.

Reply:
The pair encode functions should all take an `fr_dict_t *`.  That just needs to be passed around to where it's needed. The `fr_*_ctx_t*` arg for the encodability checks can removed.

For RADIUS, DHCPv4, DHCPv6 the functions should be identical and should probably be moved to `static inline` functions in a common header file.  The EAP-SIM-AKA encodability functions can use the common functions too, but they still need that check for bools.

Reply:
ah no, maybe that was just the decode functions.  Really the encode functions need it too.